### ISSUE TITLE
Usefulness of Pytorch-lightning for RL

### ISSUE THREAD

**Issue by JeremyAlain** [djbyrne/core_rl#21]
Hey @djbyrne 
great repository! I followed the discussion on the pytorch-lightning repo about using it for RL and found you this way. 

I was wondering, in your opinion, how useful is it to use pytorch-lightning for RL? I agree  that it makes the code "more" structured i.e. people who are used to seeing pytorch lightning, will probably be able to find things out quickly. Additionally one can still argue that things such as multi-gpu training etc. are also made easier. 

But to me, pytorch-lightning's main selling point, is the fact that you don't need to implement the whole training loop with batching etc. However when looking at your ExperienceSource class, and the EpisodeExperienceSource class, we basically still had to implement the "training" loop where an agent samples a state and then takes an action and goes to the next state. So in a sense, for supervised learning, it is very error prone, to split up your dataset into batches, and iterate over them. But here for RL it feels like we are interacting with the environment and then explicitely storing it in a dataset, just so we can then sample from it again. And for certain algorithms (that are on policy and dont have a replay buffer) this might not be necessary. And sure, lightning will do the batching for you, but in case of episodic tasks you might only use a batch size of 1 anyway. 

I really do like pytorch-lightning and I really do like your code, I am just wondering is it worth the effort? Or in other words, what are the benefits? Because somehow it feels a little bit like writing code to fit into the lightning framework. 

I am just trying to figure out for myself if I should use it for RL or not. And I thought that you might have an opinion that could convince me, because I don't see it yet. 

Cheers and thanks a lot :)

**Comment by djbyrne** [comment]
Hey Jeremy, thanks for reaching out! That’s a great question and it’s something that I have been working on for a while (and I am currently still working it out! :) )

In my opinion these are the benefits for using Lightning for RL, some of which you have already touched on. In short, the benefits for RL are the exact same benefits that other models get from PL

1: unified structure. PL is really growing in the community and my hope is that everyone will adopt a similar style/structure making research and learning easier and with higher standards.

2: distributed computing: having various types of data parallel and distributed data parallel is extremely helpful and takes away a lot of hassle
Also the TPU support

3: the other goodies: things like 16 bit training, plotting, profiling and all the other features that come out of the box

4: well maintained: PL has a really strong community and it looks like it’s here to stay

Next question, are these benefits worth the effort, or are we trying to fit RL to the framework instead of the other way around? 

My opinion is that although RL is unique in its training flow, at its core the update of the network behaves the same as with supervised learning. It still needs a batch of data to train with. This doesn't matter if its offline and coming from a randomly sampled buffer, or if its just the latest batch of steps carried out in the environment. 

Due to this, I think it actually makes sense handling these batches through the Lightning DataLoader structure. This makes it more modular and it also allows for all of lightnings extra features. 

You have a good point that for some on-policy tasks it seems like we are going out of our way to make it work by collecting it in the ExperienceSource, then putting in the dataset, then putting it through the DataLoader. There may even be a redundant buffer in there as well. In this instance, things can definitely be simplified and the middle man can be removed by having the experience source yield steps directly to the DataLoader as needed. However I am currently trying to find a structure that makes sense for both on-policy and off policy. I would prefer to no have very different structure for both.

So far I have spent the majority of my time on the off-policy methods, purely because they are easier to port to lightning due to their nature. I have only started work on the on-policy and I think there will be some changes with the whole repo before I am done.

In conclusion, I do think that Lightning is suitable for RL and I think that it is worth building RL models through Lightning for the benefits it provides. There is however, some work to be done. My hope is that with some more work and further implementations of various RL methods, we will have a clean structure for handling RL projects. 

In the mean time if you have any suggestions or ideas on the topic I would love to get your feedback :)

**Comment by JeremyAlain** [comment]
Thanks so much for your elaborate answer. This is very helpful. 

I agree that from a "theoretical" standpoint, having a dataloader that feeds data to you is still needed in RL. And I would even agree more that for algorithms such as DQN, pytorch-lightning is fitting very well. I think its just the nature of on-policy algorithms that makes it slightly weird. I will probably also try to implement something, if I have a useful idea I will let you know. 

Also I agree that the fact that you want a data-loader that generalizes to all kinds of RL-algorithms, makes them look a little more elaborate. But on the other hand you have a unified framework for all your algorithms. So maybe for an off-policy algorithm alone, I might find something more simple. 

Thanks for your reply.
